doc/man/slogverify.1.in
.TH "slogverify"1 "" "" slogverify
.PP
.SS "Name"
.PP
slogverify — Verify cryptographically secured logs.
.SS "Synopsis"
.PP
slogverify [options] [arguments]
.SS "Description"
.PP
The slogverify utility is used to verify the integrity of cryptographically secured logs and to decrypt log entries produced in a syslog-ng secure logging environment.
.RS
.IP * 2
Normal mode: slogverify -k &lt;host key file&gt; -m &lt;input MAC file&gt; &lt;input file&gt; &lt;output file&gt; [buffers]
.IP *
Iterative mode: slogverify -i -p &lt;previous host key&gt; -r &lt;previous MAC&gt; -m &lt;current MAC&gt; &lt;input file&gt; &lt;output file&gt; [buffers]
.RE
.SS "Arguments"
.RS
.IP * 2
input file
.IP
An encrypted log file from the syslog-ng secure logging environment that will be verified.
.IP *
output file
.IP
The file that will contain the plain text log entries after decryption and verification.
.IP *
buffers
.IP
Optional number of input buffers. The number of buffers can be used for performance adjustments in case the log file to be verified is very large and cannot be processed at once. It is a positive number of log entries that can be held in memory during verification. The minimum number if 10 and the maximum number is 4294967295. If this argument is not supplied the default of 1000 is used.
.RE
.SS "Options"
.RS
.IP * 2
--iterative or -i
.IP
Iterative mode. This is useful in case the log files are periodically copied from the system on which they where generated to central collector. As log rotation, i.e. overwriting log files in order to preserve space cannot be done in a secure logging environment, the iterative mode can be used instead. This works as follows: If a certain storage limit is reached the log file together with the host key and the MAC file is copied to new destination and the old file is deleted. The verification is then performed in iterations, i.e. separately for each file that was retrieved from the log host. For this to work, it is important to always retrieve the corresponding host key and MAC files. The process can be automated, e.g. by calling slogverify in iterative mode from a script.<
.IP *
--key-file or -k
.IP
The initial host key (k0). This option is used in normal mode only.
.IP *
--mac-file or -m
.IP
The current MAC file used.
.IP *
--prev-key-file or -p
.IP
The host key corresponding to the previous log file. This option can be used in iterative mode only. In theory, this can be initial host key (k0) but using this key might generate warnings, as the gap between the first log entry ever (log entry 0) and the first log entry of the current log file might be large.
.IP *
--prev-mac-file or -r
.IP
The MAC file from the previous log file. This option can only be used in iterative mode.
.IP *
--help or -h
.IP
Display a help message.
.RE
.SS "Files"
.PP
/usr/bin/slogverify
.PP
/etc/syslog-ng.conf
.SS "See also"
.PP
\fBsyslog-ng.conf.5\fP (\fIhttps://axoflow.com/docs/axosyslog-core/app-man-syslog-ng/syslog-ng.conf.5/\fP)
.PP
\fBsecure-logging.7\fP (\fIhttps://axoflow.com/docs/axosyslog-core/app-man-syslog-ng/secure-logging.7/\fP)
.SS "Getting help"
.RS
.IP * 2
The up-to-date documentation of syslog-ng is available on the \fBAxoSyslog documentation site\fP (\fIhttps://axoflow.com/docs/axosyslog-core/\fP).
.IP *
For news and notifications about syslog-ng, visit the \fBAxoflow blog\fP (\fIhttps://axoflow.com/blog/\fP).
.IP *
If you want to contact the developers directly to help with problems or report issues, contact us on \fBDiscord\fP (\fIhttps://discord.gg/583Z4wjem2\fP) or \fBGitHub\fP (\fIhttps://github.com/syslog-ng/syslog-ng/issues/\fP).
.RE
.PP
This manual page is maintained by \fBAxoflow\fP (\fIhttps://axoflow.com/\fP)
Last modified November 24, 2023: \fBMinor manpage fix (2f0a144)\fP (\fIhttps://github.com/axoflow/axosyslog-core-docs/commit/2f0a144ec391a3182da3af2f2cc136042d5608e9\fP)

NEWS
3.5.6
        Mon,  4 Aug 2014 13:39:42 +0200

        This is the sixth bug-fix release for the 3.5.x series.
        Upgrading from earlier versions is highly recommended, as the
        changes in this release are very small, yet, also very
        important for most platforms and workloads.

        Bugfixes
        ========

        * A major memory leak was fixed in the value-pairs framework,
          which affects $(format-json), MongoDB, AMQP and more. The
          leak was supposed to be fixed in 3.5.5, but due to a merging
          mistake, it was missed.

        * The configure script now detects Linux capabilities
          properly, there is no need to use --disable-linux-caps on
          non-Linux platforms anymore.

        * The pdbtool merge command will now generate version 4
          patterndb files, instead of version 3.

        Credits
        =======

        syslog-ng is developed as a community project, and as such it relies
        on volunteers to do the work necessary to produce syslog-ng.

        Reporting bugs, testing changes, writing code or simply providing
        feedback are all important contributions, so please if you are a
        user of syslog-ng, contribute.

        These people have helped in this release:

        Fabien Wernli
        Gergely Nagy
        Peter Czanik

3.5.5
        Mon, 21 Jul 2014 17:00:00 +0200

        This is the fifth bug-fix release for the 3.5.x series.
        Starting with this release, the 3.5 branch replaces 3.4 as the
        current "stable" version. Upgrading from earlier versions -
        and especially earlier branches - is highly recommended.

        New features
        ============

        * The facility() filter now supports ranges, allowing one to
          capture a range of facilities: facility(local0..local6).

        * The $(format-json) template function now supports SEQNUM.

        * The upstart script in contrib has been improved to allow the
          user to override certain settings (such as the configuration
          file, and the pidfile), and allow extra options to be
          specified.

        Bugfixes
        ========

        * Major memory leaks have been fixed in various parts of
          syslog-ng, including but not limited to network and file
          destinations and sources, value-pairs (thus affecting
          mongodb, $(format-json), and others), template handling and
          rewrite rules.

          Some of these leaks only happened during reload, some
          leaked on every message. They have now been plugged.

        * The value-pairs framework has seen a lot of updates, too:

          - It now honors time zone settings, so everything that uses
            value-pairs, will correctly take time zone settings into
            account.

          - Using globs in the key() option of value pairs will now
            also search macros for matches (thus key("D*") will match
            DATE too).

          - The syslog-proto (RFC5424) scope now correctly includes
            SDATA.

          - A bug that resulted in hierarchical elements being
            misparsed and made to appear in random places have been
            fixed.

        * The body() and subject() options of the SMTP destinations
          are now mandatory, and a double-free was also fixed in the
          driver.

        * Minor fixes were made to the $(env) template function, so
          that when looking up an empty environment variable, we will
          no longer trigger a glib warning. A misleading error message
          was also corrected.

        * The pdbtool patternize tool will now output version 4 PDB
          XML.

        * The configure script will not try to configure the AMQP
          module if no suitable version of python is installed. This
          avoids a difficult to diagnose error message during
          compilation.

        Developer news
        ==============

        * When compiled with internal ivykis, we will now install its
          headers too, along with our own. This is to make it easier
          for the Incubator to be compiled against a syslog-ng with
          embedded ivykis.

        * Plugin symbols are now loaded differently, in a way that
          lets plugins load further plugins, and use already loaded
          symbols. This is required for the various scripting
          destinations in the Incubator.

        * Starting with this release, we are installing the testing
          helpers as a convenience library (available under the
          syslog-ng-test pkg-config module). This allows third party
          projects to use our helpers.

        Credits
        =======

        syslog-ng is developed as a community project, and as such it relies
        on volunteers to do the work necessary to produce syslog-ng.

        Reporting bugs, testing changes, writing code or simply providing
        feedback are all important contributions, so please if you are a
        user of syslog-ng, contribute.

        These people have helped in this release:

        Balazs Scheidler <bazsi@balabit.hu>
        Daniel Gados
        Fabien Wernli
        Gergely Nagy <algernon@balabit.hu>
        Laszlo Budai <lbudai@balabit.hu>
        Laszlo Meszaros <lmesz@balabit.hu>
        Nick Alcock <nix@esperi.org.uk>
        Peter Czanik <czanik@balabit.hu>
        Peter Gyongyosi <gyp@balabit.hu>
        Sebastien Badia <seb@sebian.fr>
        Tibor Benke <tibor.benke@balabit.com>
        Viktor Juhasz <viktor.juhasz@balabit.com>
        Viktor Tusa <tusa@balabit.hu>
        Xufeng Zhang <xufeng.zhang@windriver.com>
        Zoltan Pallagi <pzolee@balabit.hu>

3.5.4.1
        Thu, 13 Mar 2014 10:46:07 +0100

        This is the fourth bug-fix release for the 3.5.x series, with
        a single fix for 32-bit platforms. Upgrading to this version
        is highly recommended for people using 32-bit platforms,
        others are not affected.


modules/secure-logging/slogverify/slogverify.c
/*
 * Copyright (c) 2019 Airbus Commercial Aircraft
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 * As an additional exemption you are allowed to compile & link against the
 * OpenSSL libraries as published by the OpenSSL project. See the file
 * COPYING for details.
 *
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>

#include <glib.h>

#include "messages.h"
#include "slog.h"

// Arguments and options
static gboolean iterative = FALSE;
static char *hostKey = NULL;
static char *prevHostKey = NULL;
static char *curMacFile = NULL;
static char *prevMacFile = NULL;
static char *inputLog = NULL;
static char *outputLog = NULL;
static int bufSize = DEF_BUF_SIZE;

// Return 1 on success, 0 on error
int normalMode(char *hostkey, char *MACfile, char *inputlog, char *outputlog, int bufsize)
{
  char key[KEY_LENGTH];
  guint64 counter;

  msg_info("[SLOG] INFO: Reading key file", evt_tag_str("name", hostkey));
  int ret = readKey(key, &counter, hostkey);
  if (ret!=1)
    {
      msg_error("[SLOG] ERROR: Unable to read host key", evt_tag_str("file", hostkey));
      return 0;
    }

  if (counter!=0UL)
    {
      msg_error("[SLOG] ERROR: Initial key k0 is required for verification and decryption but the supplied key read has a counter > 0.",
                evt_tag_long("Counter", counter));
      return 0;
    }

  msg_info("[SLOG] INFO: Reading MAC file", evt_tag_str("name", MACfile));
  FILE *bigMAC = fopen(MACfile, "r");
  if(bigMAC == NULL)
    {
      msg_error("[SLOG] ERROR: Unable to read MAC", evt_tag_str("file", MACfile));
      return 0;
    }

  unsigned char MAC[CMAC_LENGTH];
  if (readBigMAC(MACfile, (char *)MAC)==0)
    {
      msg_warning("[SLOG] WARNING: Unable to read MAC", evt_tag_str("file", MACfile));
    }

  FILE *input = fopen(inputlog, "r");

  if(input == NULL)
    {
      msg_error("[SLOG] ERROR: Unable to open input log", evt_tag_str("file", inputlog));
      return 0;
    }

  // Scan through file ones
  guint64 entries = 0;
  while(!feof(input))
    {
      char c = fgetc(input);
      if(c == '\n')
        {
          entries++;
        }
    }
  fclose(input);

  msg_info("[SLOG] INFO: Number of lines in file", evt_tag_long("number", entries));
  msg_info("[SLOG] INFO: Restoring and verifying log entries", evt_tag_int("buffer size", bufsize));
  ret = fileVerify((unsigned char *)key, inputlog, outputlog, MAC, entries, bufsize);

  if (ret == 0)
    {
      msg_error("[SLOG] ERROR: There is a problem with log verification. Please check log manually");
    }
  return ret;
}


// Return 1 on success, 0 on error
int iterativeMode(char *prevKey, char *prevMAC, char *curMAC, char *inputlog, char *outputlog, int bufsize)
{
  char previousKey[KEY_LENGTH];
  guint64 previousKeyCounter = 0;

  msg_info("[SLOG] INFO: Reading previous key file", evt_tag_str("name", prevKey));
  int ret = readKey(previousKey, &previousKeyCounter, prevKey);
  if (ret!=1)
    {
      msg_error("[SLOG] ERROR: Previous key could not be loaded.", evt_tag_str("file", prevKey));
      return 0;
    }

  msg_info("[SLOG] INFO: Reading previous MAC file", evt_tag_str("name", prevMAC));
  FILE *previousBigMAC = fopen(prevMAC, "r");
  if(previousBigMAC == NULL)
    {
      msg_error("[SLOG] ERROR: Unable to read previous MAC", evt_tag_str("file", prevMAC));
      return 0;
    }

  unsigned char previousMAC[CMAC_LENGTH];
  if (readBigMAC(prevMAC, (char *)previousMAC)==0)
    {
      msg_warning("[SLOG] WARNING: Unable to read previous MAC", evt_tag_str("file", prevMAC));
    }

  msg_info("[SLOG] INFO: Reading current MAC file", evt_tag_str("name", curMAC));
  FILE *currentBigMAC = fopen(curMAC, "r");
  if(currentBigMAC == NULL)
    {
      msg_error("[SLOG] ERROR: Unable to read current MAC", evt_tag_str("file", curMAC));
      return 0;
    }

  unsigned char currentMAC[CMAC_LENGTH];
  if (readBigMAC(curMAC, (char *)currentMAC)==0)
    {
      msg_warning("[SLOG] WARNING: Unable to read current MAC", evt_tag_str("file", curMAC));
    }

  FILE *input = fopen(inputlog, "r");
  if(input == NULL)
    {
      msg_error("[SLOG] ERROR: Unable to open input log", evt_tag_str("file", inputlog));
      return 0;
    }

  // Scan through file ones
  guint64 entries = 0;
  while(!feof(input))
    {
      char c = fgetc(input);
      if(c == '\n')
        {
          entries++;
        }
    }
  fclose(input);

  msg_info("[SLOG] INFO: Number of lines in file", evt_tag_long("number", entries));
  msg_info("[SLOG] INFO: Restoring and verifying log entries", evt_tag_int("buffer size", bufSize));
  ret = iterativeFileVerify(previousMAC, (unsigned char *)previousKey, inputlog, currentMAC, outputlog, entries,
                            bufsize, previousKeyCounter);

  if (ret == 0)
    {
      msg_error("[SLOG] ERROR: There is a problem with log verification. Please check log manually");
    }

  return ret;
}

// Return 0 for success and 1 on error
int main(int argc, char *argv[])
{
  int index = 0;
  SLogOptions options[] =
  {
    { "iterative", 'i', "Iterative verification", NULL, NULL },
    { "key-file", 'k', "Initial host key file", "FILE", NULL },
    { "mac-file", 'm', "Current MAC file", "FILE", NULL },
    { "prev-key-file", 'p', "Previous host key file in iterative mode", "FILE", NULL },
    { "prev-mac-file", 'r', "Previous MAC file in iterative mode", "FILE", NULL },
    { NULL }
  };

  GOptionEntry entries[] =
  {
    { options[0].longname, options[0].shortname, 0, G_OPTION_ARG_NONE, &iterative, options[0].description, NULL },
    { options[1].longname, options[1].shortname, 0, G_OPTION_ARG_CALLBACK, &validFileNameArg, options[1].description, options[1].type },
    { options[2].longname, options[2].shortname, 0, G_OPTION_ARG_CALLBACK, &validFileNameArg, options[2].description, options[2].type },
    { options[3].longname, options[3].shortname, 0, G_OPTION_ARG_CALLBACK, &validFileNameArg, options[3].description, options[3].type },
    { options[4].longname, options[4].shortname, 0, G_OPTION_ARG_CALLBACK, &validFileNameArg, options[4].description, options[4].type },
    { NULL }
  };

  GError *error = NULL;
  GOptionContext *context = g_option_context_new("INPUTLOG OUTPUTLOG [COUNTER] - Log archive verification");
  GOptionGroup *group = g_option_group_new("Basic options", "Basic log archive verification options", "basic", &options,
                                           NULL);

  g_option_group_add_entries(group, entries);
  g_option_context_set_main_group(context, group);

  if (!g_option_context_parse (context, &argc, &argv, &error))
    {
      GString *errorMsg = g_string_new(error->message);

      return slog_usage(context, group, errorMsg);
    }

  if(argc < 2 || argc > 4)
    {
      return slog_usage(context, group, NULL);
    }

  // Initialize internal messaging
  msg_init(TRUE);

  // Assign option arguments
  index = 1;
  hostKey = options[index].arg;
  index++;
  curMacFile = options[index].arg;
  index++;
  prevHostKey = options[index].arg;
  index++;
  prevMacFile = options[index].arg;
  index++;

  // Input and output file arguments
  index = 1;
  inputLog = argv[index];
  index++;
  if(!g_file_test(inputLog, G_FILE_TEST_IS_REGULAR))
    {
      GString *errorMsg = g_string_new(FILE_ERROR);
      g_string_append(errorMsg, inputLog);
      return slog_usage(context, group, errorMsg);
    }

  outputLog = argv[index];
  index++;
  if(outputLog == NULL)
    {
      return slog_usage(context, group, NULL);
    }

  // Buffer size arguments if applicable
  if (argc == 4)
    {
      bufSize = atoi(argv[index]);

      if(bufSize <= MIN_BUF_SIZE || bufSize > MAX_BUF_SIZE)
        {
          msg_error("[SLOG] ERROR: Invalid buffer size.", evt_tag_int("Size", bufSize),
                    evt_tag_int("Minimum buffer size", MIN_BUF_SIZE), evt_tag_int("Maximum buffer size", MAX_BUF_SIZE));
          g_option_context_free(context);
          return 1;
        }
    }

  int ret = 0;

  if (iterative)
    {
      if (prevHostKey == NULL || prevMacFile == NULL || curMacFile == NULL)
        {
          printf("%s", g_option_context_get_help(context, TRUE, NULL));
          g_option_context_free(context);
          return 1;
        }

      ret = 1 - iterativeMode(prevHostKey, prevMacFile, curMacFile, inputLog, outputLog, bufSize);
    }
  else
    {
      if (hostKey == NULL || curMacFile == NULL)
        {
          printf("%s", g_option_context_get_help(context, TRUE, NULL));
          g_option_context_free(context);
          return 1;
        }

      ret = 1 - normalMode(hostKey, curMacFile, inputLog, outputLog, bufSize);
    }

  // Release messaging resources
  msg_deinit();

  return ret;
}
